src/dotty/tools/dotc/repl/CompilingInterpreter.scala

```diff
@@ -0,0 +1,831 @@
+      // simple parse: just parse it, nothing else
+      def simpleParse(code: String)(implicit ctx: Context): List[Tree] = {
+        val source = new SourceFile("<console>", code.toCharArray())
+        val parser = new Parser(source)
+        val (selfDef, stats) = parser.templateStatSeq
+        stats
+      }
+      val trees = simpleParse(line)(ctx.fresh.setReporter(reporter))
+      if (reporter.hasErrors) {
+        Some(Nil) // the result did not parse, so stop
+      } else if (justNeedsMore) {
+        None
+      } else {
+        Some(trees)
+      }
+    }
+  }
+
+  /** Compile a SourceFile.  Returns the root context of the run that compiled the file.
+   */
+  def compileSources(sources: List[SourceFile])(implicit ctx: Context): Context = {
+    val reporter = newReporter
+    val run = newRun(ctx.fresh.setReporter(reporter))
+    run.compileSources(sources)
+    run.runContext
+  }
+
+  /** Compile a string.  Returns true if there are no
+   *  compilation errors, or false otherwise.
+   */
+  def compileString(code: String)(implicit ctx: Context): Boolean = {
+    val runCtx = compileSources(List(new SourceFile("<script>", code.toCharArray)))
+    !runCtx.reporter.hasErrors
+  }
+
+  override def interpret(line: String)(implicit ctx: Context): Interpreter.Result = {
+    // if (prevRequests.isEmpty)
+    //  new Run(this) // initialize the compiler // (not sure this is needed)
+    // parse
+    parse(indentCode(line)) match {
+      case None => Interpreter.Incomplete
+      case Some(Nil) => Interpreter.Error // parse error or empty input
+      case Some(tree :: Nil) if tree.isTerm && !tree.isInstanceOf[Assign] =>
+        interpret(s"val $newVarName =\n$line")
+      case Some(trees) =>
+        val req = new Request(line, newLineName)
+        if (!req.compile())
+          Interpreter.Error // an error happened during compilation, e.g. a type error
+        else {
+          val (interpreterResultString, succeeded) = req.loadAndRun()
+          if (printResults || !succeeded)
+            out.print(clean(interpreterResultString))
+          if (succeeded) {
+            prevRequests += req
+            Interpreter.Success
+          }
+          else Interpreter.Error
+        }
+    }
+  }
+
+  /** A counter used for numbering objects created by <code>bind()</code>. */
+  private var binderNum = 0
+
+  override def bind(name: String, boundType: String, value: Any)(implicit ctx: Context): Interpreter.Result = {
+    val binderName = "binder" + binderNum
+    binderNum += 1
+
+    compileString(
+      "object " + binderName +
+        "{ var value: " + boundType + " = _; " +
+        " def set(x: Any) = value=x.asInstanceOf[" + boundType + "]; }")
+
+    val binderObject =
+      Class.forName(binderName, true, classLoader)
+    val setterMethod =
+      binderObject.getDeclaredMethods.find(_.getName == "set").get
+    var argsHolder: Array[Any] = null // this roundabout approach is to try and make sure the value is boxed
```

Comment:
I'm not sure what's happening is this line and next 3 lines.
Why not simply pass `(value: Object)`